graphql_compiler/interpreter/typedefs.py

```diff
@@ -338,12 +338,140 @@ def project_neighbors(
+        a function that returns a generator instead of using a generator expression such as:
+            # BUGGY CODE FOLLOWS
+            neighbors = (
+                _your_function_that_makes_a_data_token(neighbor_id)
+                for neighbor_id in _your_function_that_gets_neighbor_ids(
+                    current_type_name, edge_info, current_token
+                )
+            )
+            # END OF BUGGY CODE
+        The above code is the victim of an unfortunate and difficult-to-find bug similar to
+        a race condition in multithreaded code:
+        - The current_token value is set to a new value in each iteration of the function's
+          outer "for data_context in data_contexts" loop.
+        - The generator in the buggy snippet refers to current_token -- but instead of referring
+          to "whatever is the value of current_token at the time the generator is defined,"
+          it instead refers to "whatever value current_token points to at each generator iteration".
+        - Since the outer loop's iterations and the buggy generator's iterations may occur in
+          an arbitrarily-iterleaved order, the buggy generator's current_token value changes
+          unpredictably during its execution therefore causing the bug.
+        - Since Python generators are not started until their first result is requested, the above
+          buggy snippet can't even guarantee that the first use of current_token has the intended
+          value -- the outer loop might have iterated before the generator was first invoked.
+
+        This bug is avoided (as in the working example code) by calling a function and supplying
+        the value of current_token to it. This creates a new Python scope (a so-called "closure")
+        that remembers the value current_token had when the closure was created. A succinct,
+        internet-search-friendly way to describe the cause of this bug is to say that in Python,
+        generators are not closures.
+
+        ## Hints supplied to this function refer to neighboring vertices
+
+        Hint kwargs such as used_property_hints, filter_hints, and neighbor_hints in this function
+        describe the desired structure of the *neighboring* vertices that this function produces,
+        instead of to the vertices supplied via the data_contexts input. For example, consider
+        the following query:
+            {
+                Foo {
+                    out_Foo_Bar {
+                        name @output(out_name: "name")
+                    }
+                }
+            }
+        The project_neighbors() call corresponding to the out_Foo_Bar edge traversal would contain
+        used_property_hints=frozenset({"name"}). This is because project_neighbors() here would
+        be called DataContexts whose current_token would point to Foo vertices, and the neighboring
+        vertices (ones along the outbound Foo_Bar edge) have their "name" property queried.
```

Comment:
I like the rewording!  To improve on the last sentence, how about this?
```
This is because used_property_hints correspond to neighboring vertices,
and the neighboring Bar vertices (along the outbound Foo_Bar edge)
are being queried for their "name" property.
```

Reply:
Queued for commit with my suggested change.